File: Tutorial4-7LayerLibrary/src/main/java/com/test/tutorial4_7layerlibrary/BaseActivity.java
package com.test.tutorial4_7layerlibrary;

import android.content.SharedPreferences;
import android.os.Bundle;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.lifecycle.ViewModelProviders;

import com.test.tutorial4_7layerlibrary.model.ToastMaker;
import com.test.tutorial4_7layerlibrary.model.User;
import com.test.tutorial4_7layerlibrary.viewmodel.BaseViewModel;

import javax.inject.Inject;

public abstract class BaseActivity extends AppCompatActivity {

    public BaseViewModel viewModel;

    @Inject
    SharedPreferences sharedPreferences;

    @Inject
    protected ToastMaker toastMaker;

    @Inject
    protected User user;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        ((BaseApplication) getApplication()).getBaseAppComponent().inject(this);

        System.out.println("BaseActivity sharedPreferences: " + sharedPreferences);
        System.out.println("BaseActivity user: " + user.getName() + ", email: " + user.getEmail());

        toastMaker.showToast("BaseActivity Toast");

        viewModel = ViewModelProviders.of(this).get(BaseViewModel.class);
    }
}

File: Tutorial4-7LayerLibrary/src/main/java/com/test/tutorial4_7layerlibrary/BaseActivity.kt
package com.test.tutorial4_7layerlibrary

import android.content.SharedPreferences
import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.ViewModelProvider
import com.test.tutorial4_7layerlibrary.model.ToastMaker
import com.test.tutorial4_7layerlibrary.viewmodel.BaseViewModel
import javax.inject.Inject

abstract class BaseActivity : AppCompatActivity() {

    private lateinit var viewModel: BaseViewModel

    @Inject
    lateinit var sharedPreferences: SharedPreferences

    @Inject
    lateinit var toastMaker: ToastMaker

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)


        println("BaseActivity sharedPreferences: $sharedPreferences")
        toastMaker.showToast("BaseActivity Toast")

        viewModel = ViewModelProvider(this).get(BaseViewModel::class.java)
    }
}